### PR TITLE
Add dynamic config change diff logging

### PR DESCRIPTION
**Pull request by Ardagan**
**What changed?**
Changed diff detection on dynamic config.

**Why?**
Fix incorrect behavior.

**How did you test it?**
local/unit testing

**Potential risks**
Dynamic config is not updated correctly.

**Is hotfix candidate?**
no

### CODE REVIEW

File: common/dynamicconfig/interface.go [hunk 1]
```diff
@@ -27,23 +27,31 @@
...
+		) (map[string]interface{}, error)
+		GetDurationValue(
+			name Key, filters map[Filter]interface{}, defaultValue time.Duration,
+		) (time.Duration, error)
+	}
+
+	FileReader interface {
+		Stat(src string) (os.FileInfo, error)
+		ReadFile(src string) ([]byte, error)
+	}
```
Comment (yiminc) [inline common/dynamicconfig/interface.go:56]: This does not need to be part of public interface. This should be internal to file based client.

File: common/dynamicconfig/constants.go [hunk 2]
```diff
@@ -641,8 +641,6 @@ const (
 	lastFilterTypeForTest
 )
 
-const DefaultNumTaskQueuePartitions = 4
```
Comment (yiminc) [inline common/dynamicconfig/constants.go:644]: why do we need to move this?

File: common/dynamicconfig/shared_constants.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dynamicconfig
+
+const DefaultNumTaskQueuePartitions = 4
```
Comment (yiminc) [inline common/dynamicconfig/shared_constants.go:27]: I would rather not create this file just for this one const.

Comment (Ardagan) [reply common/dynamicconfig/shared_constants.go:27]: I find it is worth doing to make this constant more visible.

File: common/dynamicconfig/constants.go (hunk 2 above)
Comment (Ardagan) [reply common/dynamicconfig/constants.go:644]: moved to shared_constants.go

File: common/dynamicconfig/file_based_client.go [hunk 4]
```diff
@@ -100,15 +139,15 @@ func (fc *fileBasedClient) update() error {
...
+	info, err := fc.reader.Stat(fc.config.Filepath)
 	if err != nil {
 		return fmt.Errorf("dynamic config file: %s: %w", fc.config.Filepath, err)
 	}
 	if !info.ModTime().After(fc.lastUpdatedTime) {
 		return nil
 	}
 
-	confContent, err := os.ReadFile(fc.config.Filepath)
+	confContent, err := fc.reader.ReadFile(fc.config.Filepath)
```
[the remark this answers is not in the document]
Comment (Ardagan) [reply common/dynamicconfig/file_based_client.go:150]: We don't read file if it was not touched. So we can assume that some change was actually made.

[the remark this answers is not in the document]
Comment (yiminc) [reply common/dynamicconfig/file_based_client.go:150]: didn't realize line 146, so that is good.

**Review by yycptt — COMMENTED** [review]
Sorry, I probably lack some context here. Have we considered changing the logging behavior at the dynamic config collection level instead of the client? At client level, we will need the logging logic for each client implementation.

File: common/dynamicconfig/file_based_client.go [hunk 5]
```diff
@@ -117,7 +156,94 @@ func (fc *fileBasedClient) update() error {
...
+	oldValues := fc.basicClient.getValues()
+	err = fc.storeValues(newValues)
+	newStoredValues := fc.getValues()
+
+	fc.logDiff(oldValues, newStoredValues)
+
+	return err
+}
+
+func (fc *fileBasedClient) logDiff(old configValueMap, new configValueMap) {
```
Comment (yiminc) [inline common/dynamicconfig/file_based_client.go:168]: can you move logDiff (and related methods) to basicClient so if we ever need other client type they can be shared.

**Review by yiminc — APPROVED** [review]

**Pull request merged**